polar-core/src/kb.rs

```diff
@@ -187,6 +187,73 @@ impl KnowledgeBase {
+                                ..
+                            })) = self
+                                .constants
+                                .get(member.value().as_symbol().unwrap())
+                                .map(|t| t.value())
+                            {
+                                if let Some(rule_mro) = self.mro.get(&rule_instance.tag) {
+                                    if !rule_mro.contains(instance_id) {
+                                        Ok(RuleParamMatch::False(format!("Rule specializer {} on parameter {} must be a subclass of prototype specializer {}", rule_instance.tag,index, prototype_instance.tag)))
+                                    } else if !self.param_fields_match(&prototype_instance.fields, &rule_instance.fields) {
+                                        Ok(RuleParamMatch::False(format!("Rule specializer {} on parameter {} did not match prototype specializer {} because the specializer fields did not match.", rule_instance.to_polar(), index, prototype_instance.to_polar())))
+                                    } else {
+                                        Ok(RuleParamMatch::True)
+                                    }
+                                } else {
+                                    Err(error::OperationalError::InvalidState(format!(
+                                            "All registered classes must have a registered MRO. Class {} does not have a registered MRO.",
+                                            &rule_instance.tag
+                                        )).into())
+                                }
+                            } else {
+                                unreachable!("Unregistered specializer classes should be caught before this point.");
+                            }
+                        });
+
+                        let mut success = false;
+                        for check in member_checks {
+                            match check {
+                                Ok(RuleParamMatch::True) if !success => success = true,
+                                Err(e) => return Err(e),
+                                _ => (),
+                            }
+                        }
+
+                        if !success {
+                            return Ok(RuleParamMatch::False(format!("Rule specializer {} on parameter {} must be a member of prototype specializer {}", rule_instance.tag,index, prototype_instance.tag)));
+                        }
+                    }
+                    if !self.param_fields_match(&prototype_instance.fields, &rule_instance.fields) {
+                        RuleParamMatch::False(format!("Rule specializer {} on parameter {} did not match prototype specializer {} because the specializer fields did not match.", rule_instance.to_polar(), index, prototype_instance.to_polar()))
+                    } else {
+                        RuleParamMatch::True
+                    }
```

Comment:
It seems like this could be broken into a `check_mro` function because the same logic is used in the next branch.

Reply:
302b9ede